File: sources/iOS/ios-communications/Tests/PolarBleSdkTests/PolarFirmwareUpdateUtilsTest.swift
//  Copyright © 2024 Polar. All rights reserved.

import XCTest
import RxSwift

class PolarFirmwareUpdateUtilsTest: XCTestCase {

    var mockClient: MockBlePsFtpClient!

    override func setUpWithError() throws {
        mockClient = MockBlePsFtpClient()
    }

    override func tearDownWithError() throws {
        mockClient = nil
    }

    func testReadDeviceFirmwareInfo_shouldReturnFirmwareInfo() {
        // Arrange
        let expectedDeviceId = "123456"
        let expectedFirmwareVersion = "1.2.0"
        let expectedModelName = "Model"
        let expectedHardwareCode = "00.112233"
        
        let proto = Data_PbDeviceInfo.with {
            $0.deviceVersion = .with {
                $0.major = 1
                $0.minor = 2
                $0.patch = 0
            }
            $0.modelName = expectedModelName
            $0.hardwareCode = expectedHardwareCode
        }
    
        let mockResponseData = try! proto.serializedData()
        mockClient.requestReturnValue = Single.just(mockResponseData)

        // Act
        let firmwareInfo = PolarFirmwareUpdateUtils.readDeviceFirmwareInfo(client: mockClient, deviceId: expectedDeviceId)

        // Assert
        XCTAssertEqual(firmwareInfo?.deviceFwVersion, expectedFirmwareVersion)
        XCTAssertEqual(firmwareInfo?.deviceModelName, expectedModelName)
        XCTAssertEqual(firmwareInfo?.deviceHardwareCode, expectedHardwareCode)
        XCTAssertEqual(mockClient.requestCalls.count, 1)
    }
    
    func testIsAvailableFirmwareVersionHigher_shouldReturnTrue_whenCurrentVersionIsSmallerThanAvailableVersion() {
        XCTAssertTrue(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                currentVersion: "1.0.0",
                availableVersion: "2.0.0"
            )
        )
        XCTAssertTrue(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                currentVersion: "2.0.0",
                availableVersion: "2.0.1"
            )
        )
        XCTAssertTrue(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                currentVersion: "2.0.0",
                availableVersion: "2.1.0"
            )
        )
        XCTAssertTrue(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                currentVersion: "2.0.0",
                availableVersion: "3.0.0"
            )
        )
    }
    
    func testIsAvailableFirmwareVersionHigher_shouldReturnFalse_whenCurrentVersionIsSameOrHigherThanAvailableVersion() {
        XCTAssertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                currentVersion: "2.0.0",
                availableVersion: "1.0.0"
            )
        )
        XCTAssertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                currentVersion: "2.0.1",
                availableVersion: "2.0.0"
            )
        )
        XCTAssertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                currentVersion: "2.1.0",
                availableVersion: "2.0.0"
            )
        )
        XCTAssertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                currentVersion: "3.0.0",
                availableVersion: "2.0.0"
            )
        )
        XCTAssertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                currentVersion: "2.0.0",
                availableVersion: "2.0.0"
            )
        )
    }
    
    func testFwFileComparatorSortsFilesCorrectly() {
           // Arrange
           let btFile = "BTUPDAT.BIN"
           let sysFile = "SYSUPDAT.IMG"
           let touchFile = "TCHUPDAT.BIN"
           var files = [btFile, sysFile, touchFile]
           
           // Act
           files.sort { PolarFirmwareUpdateUtils.FwFileComparator.compare($0, $1) == .orderedAscending }
           
           // Assert
           XCTAssertEqual(files[0], btFile, "First file should be BTUPDAT.BIN")
           XCTAssertEqual(files[1], touchFile, "Second file should be TCHUPDAT.BIN")
           XCTAssertEqual(files[2], sysFile, "Last file should be SYSUPDAT.IMG")
       }
       
       func testFwFileComparatorKeepsAlreadySortedFiles() {
           // Arrange
           let f1 = "BTUPDAT.BIN"
           let f2 = "TCHUPDAT.BIN"
           let f3 = "SYSUPDAT.IMG"
           var files = [f1, f2, f3]
           
           // Act
           files.sort { PolarFirmwareUpdateUtils.FwFileComparator.compare($0, $1) == .orderedAscending }
           
           // Assert
           XCTAssertEqual(files[0], f1, "Files should maintain initial order if already sorted")
           XCTAssertEqual(files[1], f2, "Files should maintain initial order if already sorted")
           XCTAssertEqual(files[2], f3, "Files should maintain initial order if already sorted")
       }
}

File: sources/Android/android-communications/library/src/test/java/com/polar/sdk/api/model/utils/PolarFirmwareUpdateUtilsTest.kt
package com.polar.sdk.api.model.utils

import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.impl.utils.PolarFirmwareUpdateUtils
import com.polar.sdk.impl.utils.PolarFirmwareUpdateUtils.FwFileComparator
import fi.polar.remote.representation.protobuf.Device
import fi.polar.remote.representation.protobuf.Structures.PbVersion
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import io.reactivex.rxjava3.core.Single
import org.junit.Assert
import org.junit.Test
import protocol.PftpRequest
import java.io.ByteArrayOutputStream
import java.io.File


class PolarFirmwareUpdateUtilsTest {

    private val mockClient = mockk<BlePsFtpClient>()
    private val firmwareFilePath = "/DEVICE.BPB"

    @Test
    fun `readDeviceFirmwareInfo() should return firmware info`() {
        // Arrange
        val deviceId = "123456"
        val firmwareVersion = "1.2.0"
        val modelName = "Model"
        val hardwareCode = "00112233.01"

        val proto = Device.PbDeviceInfo.newBuilder()
                .setDeviceVersion(PbVersion.newBuilder().setMajor(1).setMinor(2).setPatch(0))
                .setModelName(modelName)
                .setHardwareCode(hardwareCode)
                .build()

        val mockResponseContent = ByteArrayOutputStream().apply {
            proto.writeTo(this)
        }

        every { mockClient.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                        .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                        .setPath(firmwareFilePath)
                        .build().toByteArray()
        ) } returns Single.just(mockResponseContent)

        // Act
        val firmwareInfoSingle = PolarFirmwareUpdateUtils.readDeviceFirmwareInfo(mockClient, deviceId)
        val firmwareInfo = firmwareInfoSingle.blockingGet()

        // Assert
        verify {
            mockClient.request(
                    PftpRequest.PbPFtpOperation.newBuilder()
                            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                            .setPath(firmwareFilePath)
                            .build().toByteArray()
            )
        }
        confirmVerified(mockClient)

        assert(firmwareInfo.deviceFwVersion == firmwareVersion)
        assert(firmwareInfo.deviceModelName == modelName)
        assert(firmwareInfo.deviceHardwareCode == hardwareCode)
    }

    @Test
    fun `isAvailableFirmwareVersionHigher() should return true when current version is smaller than available version`() {
        Assert.assertTrue(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                "1.0.0",
                "2.0.0"
            )
        )
        Assert.assertTrue(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                "2.0.0",
                "2.0.1"
            )
        )
        Assert.assertTrue(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                "2.0.0",
                "2.1.0"
            )
        )
        Assert.assertTrue(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                "2.0.0",
                "3.0.0"
            )
        )
    }

    @Test
    fun `isAvailableFirmwareVersionHigher() should return false when current version is same or higher than available version`() {
        Assert.assertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                "2.0.0",
                "1.0.0"
            )
        )
        Assert.assertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                "2.0.1",
                "2.0.0"
            )
        )
        Assert.assertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                "2.1.0",
                "2.0.0"
            )
        )
        Assert.assertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                "3.0.0",
                "2.0.0"
            )
        )
        Assert.assertFalse(
            PolarFirmwareUpdateUtils.isAvailableFirmwareVersionHigher(
                "2.0.0",
                "2.0.0"
            )
        )
    }

    @Test
    fun `FwFileComparator sorts files correctly`() {
        // Arrange
        val btFile = mockFile("BTUPDAT.BIN")
        val sysFile = mockFile("SYSUPDAT.IMG")
        val touchFile = mockFile("TCHUPDAT.BIN")
        val files = mutableListOf(btFile, sysFile, touchFile)

        // Act
        files.sortWith(FwFileComparator())

        // Assert
        Assert.assertEquals(btFile, files[0])
        Assert.assertEquals(touchFile, files[1])
        Assert.assertEquals(sysFile, files[2])
    }

    @Test
    fun `FwFileComparator keeps already sorted files`() {
        // Arrange
        val f1 = mockFile("BTUPDAT.BIN")
        val f2 = mockFile("TCHUPDAT.BIN")
        val f3 = mockFile("SYSUPDAT.IMG")
        val files = mutableListOf(f1, f2, f3)

        // Act
        files.sortWith(FwFileComparator())

        // Assert
        Assert.assertEquals(f1, files[0])
        Assert.assertEquals(f2, files[1])
        Assert.assertEquals(f3, files[2])
    }

    private fun mockFile(name: String): File {
        val file = mockk<File>()
        every { file.name } returns name
        return file
    }
}